printing/proof_diffs.ml

```diff
@@ -628,47 +628,39 @@ let db_goal_map op np ng_to_og =
    to create the map from new goal ids to old goal ids.
 *)
 let make_goal_map op np =
```

Comment:
IINM the kernel keeps a record of all resolved evars and their values.  If so, that would make it much easier to match new and old goals without having to traverse the constrs (as in the lengthy `match_goals` routine above).  I didn't know that at time I wrote this.  That same info would also make it easy construct a tree of the derived-from relationships of the goals, which may be interesting to show to users in some form, someday.

Reply:
> the kernel keeps a record of all resolved evars and their values

It's not the kernel, it's the evarmap.

> That same info would also make it easy construct a tree of the derived-from relationships of the goals

... except that this is exactly what the Print Dependent Evar option tries to achieve, and that it does not make sense. Evars are like pointers, and the whole system is designed so that pointer dereference should be transparent. To make a pretty realistic comparison, it's like ML algebraic types which are under the hood implemented with pointers but the only thing you care about is pattern-matching over them. In particular, some pieces of the engine transparently replace evars with their expansion (e.g. when converting) so that you lose this pointer inheritance property.

Reply:
I don't fully follow your analogy; I've never thought about how to implement ML algebraic types.

If I'm eventually convinced that a completely accurate proof tree representation is truly valuable to most users, I'll look into it more, but not anytime soon. Perhaps there is a way the relationships for the problematic cases could be efficiently captured as they're generated. The current `match_goals` doesn't always succeed, but AFAIK the failures are quite infrequent so it's a good heuristic.